src/compiler/utilities.ts

```diff
@@ -2414,6 +2414,66 @@ namespace ts {
         return output;
     }
 
+    /**
+     * Serialize an object graph into a JSON string. This is intended only for use on an acyclic graph
+     * as the fallback implementation does not check for circular references by default.
+     */
+    export const stringify: (value: any) => string = JSON && JSON.stringify ? JSON.stringify : function stringify(value: any): string {
+        if (Debug.shouldAssert(AssertionLevel.Aggressive)) {
+            Debug.assert(!hasCycles(value, []), "Detected circular reference before serializing object graph.");
+        }
+
+        return value === undefined ? undefined : stringifyValue(value);
+    };
+
+    function hasCycles(value: any, stack: any[]) {
+        /* tslint:disable:no-null */
+        if (typeof value !== "object" || value === null) {
+            return false;
+        }
+        /* tslint:enable:no-null */
+
+        if (stack.lastIndexOf(value) !== -1) {
+            return true;
+        }
+
+        stack.push(value);
+
+        for (const key in value) {
+            if (hasProperty(value, key) && hasCycles(value[key], stack)) {
+                return true;
+            }
+        }
+
+        stack.pop();
+        return false;
+    }
+
+    function stringifyValue(value: any): string {
+        return typeof value === "string" ? `"${escapeString(value)}"`
+             : typeof value === "number" ? isFinite(value) ? String(value) : "null"
+             : typeof value === "boolean" ? value ? "true" : "false"
+             : typeof value === "object" ? isArray(value) ? stringifyArray(value) : stringifyObject(value)
+             : /*fallback*/ "null";
+    }
+
+    function stringifyArray(value: any) {
+        return `[${reduceLeft(value, stringifyElement, "")}]`;

```

Comment:
Lately I've been very allocation-averse and that approach creates two temporary arrays.
________________________________
From: Nathan Shively-Sanders<mailto:notifications@github.com>
Sent: ‎11/‎25/‎2015 10:12 AM
To: Microsoft/TypeScript<mailto:TypeScript@noreply.github.com>
Cc: Ron Buckton<mailto:Ron.Buckton@microsoft.com>
Subject: Re: [TypeScript] Extract source map generation logic out of the emitter. (#5780)


In src/compiler/utilities.ts<https://github.com/Microsoft/TypeScript/pull/5780#discussion_r45900346>:

> +        }
> +
> +        stack.pop();
> +        return false;
> +    }
> +
> +    function stringifyValue(value: any): string {
> +        return typeof value === "string" ? `"${escapeString(value)}"`
> +             : typeof value === "number" ? isFinite(value) ? String(value) : "null"
> +             : typeof value === "boolean" ? value ? "true" : "false"
> +             : typeof value === "object" ? isArray(value) ? stringifyArray(value) : stringifyObject(value)
> +             : /*fallback*/ "null";
> +    }
> +
> +    function stringifyArray(value: any) {
> +        return `[${reduceLeft(value, stringifyElement, "")}]`;


why not map(filter(value, v => v), stringifyValue).join(',') ?

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/TypeScript/pull/5780/files#r45900346>.